### PR TITLE
printf: Move implementation inside VVL

### PR DESCRIPTION
**Pull request by spencer-lunarg**
This was brought over from the current implementations in SPIRV-Tools (`source/opt/inst_debug_printf_pass.cpp`) for 3 reasons (after lots of discussions)

1. There are bugs that need be fixed and easier to iterate without having to grab a dependency
2. To get GPU-AV and work with DebugPrintf we need a more unified way the shaders are instrumented
3. Allows future maintainer to not need to learn yet-another-code-base to make changes

There is more I want to fix/add but this is the initial "get it to work the same" change... there are limitations (as noted from disabled tests) of the old implementation and plan to fix those once this is in

### CODE REVIEW

File: layers/gpu/spirv/debug_printf_pass.cpp [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+        check_block->CreateInstruction(spv::OpULessThanEqual, {bool_type_id, less_than_equal_id, int_add_id, array_length_id});
+
+        const uint32_t merge_block_label_id = merge_block->GetLabelId();
+        check_block->CreateInstruction(spv::OpSelectionMerge, {merge_block_label_id, spv::SelectionControlMaskNone});
+
+        const uint32_t store_block_label_id = store_block->GetLabelId();
+        check_block->CreateInstruction(spv::OpBranchConditional, {less_than_equal_id, store_block_label_id, merge_block_label_id});
+    }
+
+    // Store how many bytes
```
Comment (jeremyg-lunarg) [inline layers/gpu/spirv/debug_printf_pass.cpp:338]: Is this really bytes or is it 32 bit words?

File: layers/gpu/spirv/debug_printf_pass.cpp [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+
+        const uint32_t access_chain_id = module_.TakeNextId();
+        store_block->CreateInstruction(spv::OpAccessChain,
+                                       {pointer_type_id, access_chain_id, output_buffer_variable_id_, one_id, int_add_id});
+
+        const uint32_t shader_id = module_.type_manager_.GetConstantUInt32(module_.shader_id_).Id();
+        store_block->CreateInstruction(spv::OpStore, {access_chain_id, shader_id});
+    }
+
+    // Write a byte for each argument
```
Comment (jeremyg-lunarg) [inline layers/gpu/spirv/debug_printf_pass.cpp:363]: bytes?

**Review by jeremyg-lunarg — APPROVED** [review]
LGTM this looks like what I remember of the spirv-tools implementation but simpler.

File: layers/gpu/spirv/debug_printf_pass.cpp (hunk 1 above)
Comment (spencer-lunarg) [reply layers/gpu/spirv/debug_printf_pass.cpp:338]: yes, how many 32-bit words, will update

**Comment by arno-lunarg** [comment]
Will review tomorrow, sorry I did not take the time this week (￣(工)￣)

File: layers/gpu/spirv/function_basic_block.cpp [hunk 3]
```diff
@@ -117,6 +117,13 @@ BasicBlockIt Function::InsertNewBlock(BasicBlockIt it) {
     return new_block_it;
 }
 
+void Function::InitBlocks(uint32_t count) {
+    for (uint32_t i = 0; i < count; i++) {
```
Comment (arno-lunarg) [inline layers/gpu/spirv/function_basic_block.cpp:121]:
```suggestion
    blocks_.reserve(blocks_.size() + count);
    for (uint32_t i = 0; i < count; i++) {
```

File: layers/gpu/spirv/module.cpp [hunk 4]
```diff
@@ -227,28 +226,68 @@ void Module::AddExtension(const char* extension) {
...
+    debug_name_.emplace_back(std::move(new_inst));
+}
+
+void Module::AddDecoration(uint32_t target_id, spv::Decoration decoration, const std::vector<uint32_t>& operands) {
+    auto new_inst = std::make_unique<Instruction>((uint32_t)(operands.size() + 3), spv::OpDecorate);
+    new_inst->Fill({target_id, (uint32_t)decoration});
+    if (!operands.empty()) {
+        new_inst->Fill(operands);
+    }
+    annotations_.push_back(std::move(new_inst));
```
Comment (arno-lunarg) [inline layers/gpu/spirv/module.cpp:243]: Nit:
```suggestion
    annotations_.emplace_back(std::move(new_inst));
```

**Review by arno-lunarg — APPROVED** [review]
LGTM! Thank you for doing this, must have been as tedious as it was needed!

File: layers/gpu/spirv/function_basic_block.cpp (hunk 3 above)
Comment (spencer-lunarg) [reply layers/gpu/spirv/function_basic_block.cpp:121]: I need to go around and see if we can do this reserve more often for SPIR-V stuff

File: layers/gpu/spirv/bindless_descriptor_pass.cpp [hunk 5]
```diff
@@ -424,5 +425,9 @@ bool BindlessDescriptorPass::AnalyzeInstruction(const Function& function, const
     return true;
 }
 
+void BindlessDescriptorPass::PrintDebugInfo() {
+    std::cout << "BindlessDescriptorPass\n\tinstrumentation count: " << instrumented_count_ << "\n";
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/spirv/bindless_descriptor_pass.cpp:429]: going to do a big find-and-replace after, we do this a lot actually

File: layers/gpu/debug_printf/debug_printf.cpp [hunk 6]
```diff
@@ -80,48 +83,66 @@ void Validator::DestroyBuffer(BufferInfo &buffer_info) {
...
                                  std::vector<uint32_t> &out_instrumented_spirv) {
     if (input[0] != spv::MagicNumber) return false;
 
-    // Load original shader SPIR-V
-    out_instrumented_spirv.clear();
-    out_instrumented_spirv.reserve(input.size());
-    out_instrumented_spirv.insert(out_instrumented_spirv.end(), &input.front(), &input.back() + 1);
+    std::vector<uint32_t> binary;
+    binary.reserve(input.size());
+    binary.insert(binary.end(), &input.front(), &input.back() + 1);
```
[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/debug_printf/debug_printf.cpp:108]: I know before we had an issue with how `spirv-opt` needed the input to be a certain type (I think `std::vector<uint32_t>`) but now that this is all gone, I need to look into making sure there is no copy all the way down the chain from when we get the SPIR-V

[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu/debug_printf/debug_printf.cpp:108]: idk if we need a copy or not anymore neither, I **think** the issue was before we might have touched that vector, but now I consume things in its own representation, so it **should** be a const ref... need to make sure (here and normal GPU-AV)

[the remark this answers is not in the document]
Comment (arno-lunarg) [reply layers/gpu/debug_printf/debug_printf.cpp]: worth looking into!

**Pull request merged**